internal/stage_h6.go

```diff
@@ -0,0 +1,134 @@
+// Stage 5: Down-arrow navigation test
+// This test checks if the shell supports traversing through history with both up and down arrow keys.
+// It sends a few commands, simulates up and down arrow presses, and checks if the correct command is recalled at the prompt.
+
+package internal
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/codecrafters-io/shell-tester/internal/assertions"
+	"github.com/codecrafters-io/shell-tester/internal/logged_shell_asserter"
+	"github.com/codecrafters-io/shell-tester/internal/shell_executable"
+	"github.com/codecrafters-io/shell-tester/internal/test_cases"
+	"github.com/codecrafters-io/tester-utils/random"
+	"github.com/codecrafters-io/tester-utils/test_case_harness"
+)
+
+func testH6(stageHarness *test_case_harness.TestCaseHarness) error {
+	shell := shell_executable.NewShellExecutable(stageHarness)
+	asserter := logged_shell_asserter.NewLoggedShellAsserter(shell)
+
+	if err := asserter.StartShellAndAssertPrompt(true); err != nil {
+		return err
+	}
+
+	upArrow := "\x1b[A"
+	downArrow := "\x1b[B"
+
+	// Execute initial commands using CommandResponseTestCase
+	randomWords1 := strings.Join(random.RandomWords(2), " ")
+	randomWords2 := strings.Join(random.RandomWords(2), " ")
+	randomWords3 := strings.Join(random.RandomWords(2), " ")
+	randomCommand := getRandomInvalidCommand()
+
+	commandTestCases := []test_cases.CommandResponseTestCase{
+		{Command: "echo " + randomWords1, ExpectedOutput: randomWords1, SuccessMessage: commandSuccessMessage},
+		{Command: "echo " + randomWords2, ExpectedOutput: randomWords2, SuccessMessage: commandSuccessMessage},
+	}
+
+	for _, command := range commandTestCases {
+		if err := command.Run(asserter, shell, stageHarness.Logger); err != nil {
+			return err
+		}
+	}
+
+	// Run invalid command test case
+	invalidCommandTestCase := test_cases.InvalidCommandTestCase{
+		Command: randomCommand,
+	}
+	if err := invalidCommandTestCase.Run(asserter, shell, stageHarness.Logger); err != nil {
+		return err
+	}
+
+	// Run the last echo command
+	echoCommand := test_cases.CommandResponseTestCase{
+		Command:        "echo " + randomWords3,
+		ExpectedOutput: randomWords3,
+		SuccessMessage: commandSuccessMessage,
+	}
+	if err := echoCommand.Run(asserter, shell, stageHarness.Logger); err != nil {
+		return err
+	}
+
+	// Test up-arrow navigation (going back in history)
+	expectedCommands := []struct {
+		command string
+		message string
+	}{
+		{echoCommand.Command, "echo " + randomWords3},
+		{randomCommand, randomCommand},
+	}
+
+	for _, expected := range expectedCommands {
+		if err := shell.SendCommandRaw(upArrow); err != nil {
+			return err
+		}
+		stageHarness.Logger.Infof("Pressed %q (expecting to recall %q)", "<UP ARROW>", expected.message)
+		asserter.AddAssertion(assertions.SingleLineAssertion{
+			ExpectedOutput: "$ " + expected.command,
+			FallbackPatterns: []*regexp.Regexp{
+				regexp.MustCompile(`^\s*` + expected.command + `\s*$`),
```

Comment:
Let's investigate why this is required - using alpine on my system I don't see any weird formatting